### PR TITLE
Monad[DynamoRead]

### PR DESCRIPTION
**Pull request by amrhassan**

### CODE REVIEW

**Comment by amrhassan** [comment]
`Monad[DynamoRead].tailRecM` is not actually tail-recursive.

Re-implementing this seems like unnecessary effort, I'd suggest going through with re-designing this typeclass as `Kleisli[DynamoType, Either[DynamoReadResult, ?], ?`, then you get all the Cats instances for free.

File: src/main/scala/dynamo/ast/reads/DynamoRead.scala [hunk 1]
```diff
@@ -11,15 +11,23 @@ trait DynamoRead[A] { self =>
...
+    def pure[A](x: A): DynamoRead[A] = DynamoRead[A] { _ => DynamoReadSuccess(x) }
+    def flatMap[A, B](fa: DynamoRead[A])(f: (A) => DynamoRead[B]): DynamoRead[B] =
+      DynamoRead { x =>
+        fa.read(x) match {
+          case DynamoReadSuccess(a) => f(a).read(x)
+          case DynamoReadError(path, err) => DynamoReadError(path, err)
+        }
+      }
+    def tailRecM[A, B](a: A)(f: (A) => DynamoRead[Either[A, B]]): DynamoRead[B] =
+      DynamoRead { x =>
```
Comment (louis-forite) [inline src/main/scala/dynamo/ast/reads/DynamoRead.scala]: neat. I don't know how I missed to work in the context of a DynamoRead. It it missing however the type annotation `DynamoRead[B]`

File: src/main/scala/dynamo/ast/reads/DynamoRead.scala [hunk 2]
```diff
@@ -11,15 +11,23 @@ trait DynamoRead[A] { self =>
...
-    override def pure[A](x: A): DynamoRead[A] = DynamoRead[A] { _ => DynamoReadSuccess(x) }
-
-    override def ap[A, B](ff: DynamoRead[(A) => B])(fa: DynamoRead[A]): DynamoRead[B] = DynamoRead[B] { dynamoType =>
-      fa.read(dynamoType) match {
-        case DynamoReadSuccess(a) => ff.read(dynamoType).map((aToB) => aToB(a))
-        case e:DynamoReadError => e
+  implicit val monadDynamoRead: Monad[DynamoRead] = new Monad[DynamoRead] {
+    def pure[A](x: A): DynamoRead[A] = DynamoRead[A] { _ => DynamoReadSuccess(x) }
+    def flatMap[A, B](fa: DynamoRead[A])(f: (A) => DynamoRead[B]): DynamoRead[B] =
+      DynamoRead { x =>
```
Comment (louis-forite) [inline src/main/scala/dynamo/ast/reads/DynamoRead.scala]: `DynamoRead` should be annotated `DynamoRead[B]`

**Comment by lforite** [comment]
@amrhassan Actually I noticed that I don't need a `Monad` instance there ! Thus closing the issue ! Thanks for the contribution :)

**Pull request closed without merging**